commit 19648694faaf973e7b4b0de1dbe49710e14a8ce7
Author: Simon Mellerin <simon.mellerin@makina-corpus.com>
Date:   Fri Dec 22 17:22:41 2023 +0100

    Yunohost ldap

diff --git a/config/packages/security.yaml b/config/packages/security.yaml
index 6c4457f1..ea1f3dc9 100644
--- a/config/packages/security.yaml
+++ b/config/packages/security.yaml
@@ -11,6 +11,8 @@ security:
             entity:
                 class: App\Entity\User
                 property: username
+        user_provider_yunohost:
+            id: yunohost.provider.ldap
     firewalls:
         dev:
             pattern: ^/(_(profiler|wdt)|css|images|js)/
@@ -23,6 +25,10 @@ security:
                 login_path: security.login
                 check_path: security.login
                 enable_csrf: true
+            http_basic_ldap:
+                provider: user_provider_yunohost
+                service: yunohost.ldap
+                dn_string: "uid={username},ou=users,dc=yunohost,dc=org"
             logout:
                 path: security.logout
                 target: security.login
diff --git a/config/services.yaml b/config/services.yaml
index 4410bfc5..71b9ba86 100644
--- a/config/services.yaml
+++ b/config/services.yaml
@@ -105,3 +105,16 @@ services:

     Aeneria\GrdfAdictApi\Client\GrdfAdictClientInterface:
         alias: Aeneria\GrdfAdictApi\Client\GrdfAdictClient
+
+    yunohost.provider.ldap:
+        class: App\Security\YunohostLdapUserProvider
+        arguments: ["@yunohost.ldap", "ou=users,dc=yunohost,dc=org"]
+
+    yunohost.ldap:
+        class: Symfony\Component\Ldap\Ldap
+        arguments: ['@yunohost.ldap.adapter']
+
+    yunohost.ldap.adapter:
+        class: Symfony\Component\Ldap\Adapter\ExtLdap\Adapter
+        arguments:
+            - host: "localhost"
diff --git a/src/Security/YunohostLdapUserProvider.php b/src/Security/YunohostLdapUserProvider.php
new file mode 100644
index 00000000..09ad20c6
--- /dev/null
+++ b/src/Security/YunohostLdapUserProvider.php
@@ -0,0 +1,102 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Security;
+
+use App\Entity\User;
+use App\Repository\UserRepository;
+use Symfony\Component\Ldap\Entry;
+use Symfony\Component\Ldap\Exception\ConnectionException;
+use Symfony\Component\Ldap\LdapInterface;
+use Symfony\Component\Security\Core\Exception\InvalidArgumentException;
+use Symfony\Component\Security\Core\Exception\UnsupportedUserException;
+use Symfony\Component\Security\Core\Exception\UserNotFoundException;
+use Symfony\Component\Security\Core\User\UserInterface;
+use Symfony\Component\Security\Core\User\UserProviderInterface;
+
+/**
+ * Adapted from LdapUserProvider.
+ *
+ */
+class YunohostLdapUserProvider implements UserProviderInterface
+{
+    private string $defaultSearch = '(uid={username})';
+
+    public function __construct(
+        private LdapInterface $ldap,
+        private string $baseDn,
+        private UserRepository $userRepository,
+        private ?string $searchDn = null,
+        private ?string $searchPassword = null,
+    ) {}
+
+    /**
+     * {@inheritdoc}
+     */
+    public function loadUserByUsername(string $username)
+    {
+        trigger_deprecation('symfony/security-core', '5.3', 'Method "%s()" is deprecated, use loadUserByIdentifier() instead.', __METHOD__);
+
+        return $this->loadUserByIdentifier($username);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function loadUserByIdentifier(string $identifier): UserInterface
+    {
+        try {
+            $this->ldap->bind($this->searchDn, $this->searchPassword);
+            $username = $this->ldap->escape($identifier, '', LdapInterface::ESCAPE_FILTER);
+            $query = str_replace('{username}', $identifier, $this->defaultSearch);
+            $search = $this->ldap->query($this->baseDn, $query);
+        } catch (ConnectionException $e) {
+            throw new UserNotFoundException(sprintf('User "%s" not found.', $identifier), 0, $e);
+        }
+
+        $entries = $search->execute();
+        $count = \count($entries);
+
+        if ($count > 1) {
+            throw new UserNotFoundException('More than one user found.');
+        }
+
+        $entry = $entries[0];
+
+        $identifier = $this->getAttributeValue($entry, 'mail');
+
+        return $this->userRepository->findOneBy(['username' => $identifier]);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function refreshUser(UserInterface $user)
+    {
+        if (!$user instanceof User) {
+            throw new UnsupportedUserException(sprintf('Instances of "%s" are not supported.', \get_class($user)));
+        }
+
+        return $this->userRepository->findOneBy(['username' => $user->getUsername()]);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supportsClass(string $class)
+    {
+        return User::class === $class;
+    }
+
+    private function getAttributeValue(Entry $entry, string $attribute)
+    {
+        if (!$entry->hasAttribute($attribute)) {
+            throw new InvalidArgumentException(sprintf('Missing attribute "%s" for user "%s".', $attribute, $entry->getDn()));
+        }
+
+        $values = $entry->getAttribute($attribute);
+
+        return $values[0];
+    }
+}
